pkg/controller/beat/common/common.go

```diff
@@ -0,0 +1,86 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package common
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+
+	beatv1beta1 "github.com/elastic/cloud-on-k8s/pkg/apis/beat/v1beta1"
+	commonassociation "github.com/elastic/cloud-on-k8s/pkg/controller/common/association"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/container"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/defaults"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/reconciler"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/settings"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+)
+
+type Type string
+
+type Driver interface {
+	Reconcile() *reconciler.Results
+}
+
+type DriverParams struct {
+	Client  k8s.Client
+	Context context.Context
+	Logger  logr.Logger
+
+	Beat beatv1beta1.Beat
+}
+
+func (dp *DriverParams) GetPodTemplate() corev1.PodTemplateSpec {
+	spec := dp.Beat.Spec
+	switch {
+	case spec.DaemonSet != nil:
+		return spec.DaemonSet.PodTemplate
+	case spec.Deployment != nil:
+		return spec.Deployment.PodTemplate
+	}
+
+	return corev1.PodTemplateSpec{}
+}
+
+func (dp *DriverParams) Validate() error {
```

Comment:
it feels weird to me to have this function attached to `DriverParams`. I think I would make it a standalone function: `func ValidateBeatSpec(b Beat) error`